client/my-sites/site-settings/publishing-tools/hooks/use-switch-post-by-voice-mutation.tsx

```diff
@@ -0,0 +1,63 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import wpcom from 'calypso/lib/wp';
+import { getCachePostByVoiceKey } from 'calypso/my-sites/site-settings/publishing-tools/hooks/get-cache-post-by-voice-key';
+import { getPostByVoicePath } from 'calypso/my-sites/site-settings/publishing-tools/hooks/get-post-by-voice-path';
+import { parsePostByVoiceResponse } from 'calypso/my-sites/site-settings/publishing-tools/hooks/parse-post-by-voice-response';
+import type { PostByVoice } from 'calypso/my-sites/site-settings/publishing-tools/types/post-by-voice';
+import type { PostByVoiceResponse } from 'calypso/my-sites/site-settings/publishing-tools/types/post-by-voice-response';
+
+export const useSwitchPostByVoiceMutation = ( siteId: number | null ) => {
+	const queryClient = useQueryClient();
+	const queryKey = getCachePostByVoiceKey( siteId );
+
+	return useMutation( {
+		mutationFn: async ( value: boolean ) => {
+			let response: PostByVoiceResponse;
+			if ( value ) {
+				response = await await wpcom.req.post( {
+					method: 'POST',
+					path: getPostByVoicePath( siteId ),
+					apiNamespace: 'wpcom/v2',
+				} );
+			} else {
+				response = await await wpcom.req.del( {
+					method: 'DELETE',
+					path: getPostByVoicePath( siteId ),
+					apiNamespace: 'wpcom/v2',
+				} );
+			}
+
+			return parsePostByVoiceResponse( response );
+		},
+		onMutate: async ( value: boolean ) => {
+			await queryClient.cancelQueries( {
+				queryKey,
+			} );
+
+			const restoreData = queryClient.getQueryData< PostByVoice >( queryKey );
+
+			queryClient.setQueryData< PostByVoice >( queryKey, () => {
+				if ( value ) {
+					return {
+						isEnabled: true,
+					};
+				}
+
+				return {
+					isEnabled: false,
+					code: undefined,
+				};
+			} );
```

Comment:
```suggestion
			queryClient.setQueryData< PostByVoice >( queryKey, () => {
				return {
					isEnabled: value,
					code: undefined,
				};
			} );
```
Not a requirement, but doesn't this have the same effect?

Reply:
I decided that this shortcut could be a bit confusing, due to setting `code: undefined` when we are waiting for the code from backend. But actually there is no big difference, I agree, let's go with this short version 👍